From 1947743a722394774bc38a99282e5b98dbc67614 Mon Sep 17 00:00:00 2001
From: yutingfang <yutingfang@google.com>
Date: Tue, 12 Aug 2025 16:48:55 -0700
Subject: [PATCH 14/55] Use ParceledListSlice to paginate response from
 getPackagesForOpsForDevice binder API

In a previous similar security bug fix ag/31434165 we imposed a
threshold on the number of attributed op entries returned for a single
package. But in this bug, the POC app created attributed op entries for
hundreds of packages which caused TransactionTooLargeException in getPackagesForOpsForDevice binder API.
So the fix is to add pagination in the binder response by using
ParceledListSlice.

Bug: 416490321
Test: Use the POC app in the bug report, verified after the fix there
was no more crashes from SystemUi and the device could boot
successfully.
Flag: EXEMPT security bug fix

Change-Id: If4590160d1d743f58c1aed3f923d3668a8936c57
---
 core/java/android/app/AppOpsManager.java               |  9 +++++----
 core/java/com/android/internal/app/IAppOpsService.aidl |  4 ++--
 .../java/com/android/server/appop/AppOpsService.java   | 10 +++++++---
 3 files changed, 14 insertions(+), 9 deletions(-)

diff --git a/core/java/android/app/AppOpsManager.java b/core/java/android/app/AppOpsManager.java
index ec439a228..09b651b96 100644
--- a/core/java/android/app/AppOpsManager.java
+++ b/core/java/android/app/AppOpsManager.java
@@ -8440,13 +8440,13 @@ public class AppOpsManager {
         } else {
             opCodes = null;
         }
-        final List<AppOpsManager.PackageOps> result;
         try {
-            result = mService.getPackagesForOpsForDevice(opCodes, persistentDeviceId);
+            ParceledListSlice<PackageOps> packageOps = mService.getPackagesForOpsForDevice(opCodes,
+                    persistentDeviceId);
+            return packageOps == null ? Collections.emptyList() : packageOps.getList();
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
         }
-        return (result != null) ? result : Collections.emptyList();
     }
 
     /**
@@ -8465,8 +8465,9 @@ public class AppOpsManager {
     @UnsupportedAppUsage
     public List<AppOpsManager.PackageOps> getPackagesForOps(int[] ops) {
         try {
-            return mService.getPackagesForOpsForDevice(ops,
+            ParceledListSlice<PackageOps> packageOps = mService.getPackagesForOpsForDevice(ops,
                     VirtualDeviceManager.PERSISTENT_DEVICE_ID_DEFAULT);
+            return packageOps == null ? null : packageOps.getList();
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
         }
diff --git a/core/java/com/android/internal/app/IAppOpsService.aidl b/core/java/com/android/internal/app/IAppOpsService.aidl
index f01aa80fa..ab973a3ea 100644
--- a/core/java/com/android/internal/app/IAppOpsService.aidl
+++ b/core/java/com/android/internal/app/IAppOpsService.aidl
@@ -162,6 +162,6 @@ interface IAppOpsService {
             int attributionFlags, int attributionChainId);
     void finishOperationForDevice(IBinder clientId, int code, int uid, String packageName,
             @nullable String attributionTag, int virtualDeviceId);
-   List<AppOpsManager.PackageOps> getPackagesForOpsForDevice(in int[] ops, String persistentDeviceId);
-   oneway void noteOperationsInBatch(in Map batchedNoteOps);
+    ParceledListSlice<AppOpsManager.PackageOps> getPackagesForOpsForDevice(in int[] ops, String persistentDeviceId);
+    oneway void noteOperationsInBatch(in Map batchedNoteOps);
 }
diff --git a/services/core/java/com/android/server/appop/AppOpsService.java b/services/core/java/com/android/server/appop/AppOpsService.java
index 0f6025b5a..c749d45b5 100644
--- a/services/core/java/com/android/server/appop/AppOpsService.java
+++ b/services/core/java/com/android/server/appop/AppOpsService.java
@@ -110,6 +110,7 @@ import android.content.IntentFilter;
 import android.content.pm.PackageInfo;
 import android.content.pm.PackageManager;
 import android.content.pm.PackageManagerInternal;
+import android.content.pm.ParceledListSlice;
 import android.content.pm.PermissionInfo;
 import android.content.pm.UserInfo;
 import android.database.ContentObserver;
@@ -204,6 +205,7 @@ import java.io.FileOutputStream;
 import java.io.FileWriter;
 import java.io.IOException;
 import java.io.PrintWriter;
+import java.io.StringWriter;
 import java.text.SimpleDateFormat;
 import java.time.Duration;
 import java.time.Instant;
@@ -1819,11 +1821,13 @@ public class AppOpsService extends IAppOpsService.Stub {
 
     @Override
     public List<AppOpsManager.PackageOps> getPackagesForOps(int[] ops) {
-        return getPackagesForOpsForDevice(ops, PERSISTENT_DEVICE_ID_DEFAULT);
+        ParceledListSlice<AppOpsManager.PackageOps> packageOps = getPackagesForOpsForDevice(ops,
+                PERSISTENT_DEVICE_ID_DEFAULT);
+        return packageOps == null ? null : packageOps.getList();
     }
 
     @Override
-    public List<AppOpsManager.PackageOps> getPackagesForOpsForDevice(int[] ops,
+    public ParceledListSlice<AppOpsManager.PackageOps> getPackagesForOpsForDevice(int[] ops,
             @NonNull String persistentDeviceId) {
         final int callingUid = Binder.getCallingUid();
         final boolean hasAllPackageAccess = mContext.checkPermission(
@@ -1860,7 +1864,7 @@ public class AppOpsService extends IAppOpsService.Stub {
                 }
             }
         }
-        return res;
+        return res == null ? null : new ParceledListSlice<>(res);
     }
 
     @Override
-- 
2.52.0

